README.md
# IF708-Interpreter

## Build

```bash
stack build
```

## Exec

```bash
stack exec IF708-Interpreter-exe < ./test/sourceCode.txt
```

## Test

```bash
stack test
```


ChangeLog.md
# Changelog for IF708-Interpreter

## Unreleased changes


test/Spec.hs
import Control.Exception (evaluate)
import qualified Data.Map.Strict as Map

import Test.Hspec
import Test.QuickCheck

import AST
import Lib
import Interpreter
import Parser

sourceCode :: String
sourceCode = "declare x = 150 in\n declare y = 200 in\n {while x > 0 do { x:=x-1; y:=y-1 };\n print y\n }\ndeclare x = 150 in\n declare y = 200 in\n {while x > 0 do { x:=x-1; y:=y-1 };\n print y\n }\n"

s :: Com
s = Declare "x" (Constant 150) $
    Declare "y" (Constant 200) $
    Seq
        (While
            (Greater (Variable "x") (Constant 0))
            (Seq
                (Assign "x" (Minus (Variable "x") (Constant 1)))
                (Assign "y" (Minus (Variable "y") (Constant 1)))
            )
        )
        (Print (Variable "y"))

main :: IO ()
main = hspec $ do
    describe "Smoke Test" $ do
        it "parsers source code" $
            let Just (ast, _) = apply com sourceCode
             in ast `shouldBe` s
        it "parsers source codes" $
            let as = applyAll com sourceCode
             in as `shouldBe` [s, s]
        it "evaluates ast" $
            let Just (ast, _) = apply com sourceCode
                (_, out, _) = interpret (exec ast) Map.empty
             in out `shouldBe` "50\n"
        it "respects arithmetic precedence" $
            let Just (ast, _) = apply com "print (-6 + 4 * 2 ** 3 / 5 - 7)"
                (_, out, _) = interpret (exec ast) Map.empty
            in out `shouldBe` "-7\n"
        it "respects relational precedence" $
            let Just (ast, _) = apply com "print (-6 + 4 * 2 <= 3 / 5 - 7)"
                (_, out, _) = interpret (exec ast) Map.empty
            in out `shouldBe` "0\n"
    describe "Interpreter" $ do
        it "negates" $
            let Just (ast, _) = apply com "print -1"
                (_, out, _) = interpret (exec ast) Map.empty
             in out `shouldBe` "-1\n"
        it "adds" $
            let Just (ast, _) = apply com "print (2 + 3)"
                (_, out, _) = interpret (exec ast) Map.empty
             in out `shouldBe` "5\n"
        it "subs" $
            let Just (ast, _) = apply com "print (2 - 3)"
                (_, out, _) = interpret (exec ast) Map.empty
             in out `shouldBe` "-1\n"
        it "mults" $
            let Just (ast, _) = apply com "print (2 * 3)"
                (_, out, _) = interpret (exec ast) Map.empty
             in out `shouldBe` "6\n"
        it "divs" $
            let Just (ast, _) = apply com "print (2 / 3)"
                (_, out, _) = interpret (exec ast) Map.empty
             in out `shouldBe` "0\n"
        it "pows" $
            let Just (ast, _) = apply com "print (2 ** 3)"
                (_, out, _) = interpret (exec ast) Map.empty
             in out `shouldBe` "8\n"
        it "complements" $
            let Just (ast, _) = apply com "print ~0"
                (_, out, _) = interpret (exec ast) Map.empty
             in out `shouldBe` "-1\n"
        it "ands" $
            let Just (ast, _) = apply com "print (2 & 3)"
                (_, out, _) = interpret (exec ast) Map.empty
             in out `shouldBe` "2\n"
        it "ors" $
            let Just (ast, _) = apply com "print (2 | 3)"
                (_, out, _) = interpret (exec ast) Map.empty
             in out `shouldBe` "3\n"
        it "xors" $
            let Just (ast, _) = apply com "print (2 ^ 3)"
                (_, out, _) = interpret (exec ast) Map.empty
             in out `shouldBe` "1\n"
        it "eqs" $
            let Just (ast, _) = apply com "print !0"
                (_, out, _) = interpret (exec ast) Map.empty
             in out `shouldBe` "1\n"
        it "eqs" $
            let Just (ast, _) = apply com "print (1 = 1)"
                (_, out, _) = interpret (exec ast) Map.empty
             in out `shouldBe` "1\n"
        it "nes" $
            let Just (ast, _) = apply com "print (1 != 1)"
                (_, out, _) = interpret (exec ast) Map.empty
             in out `shouldBe` "0\n"
        it "lts" $
            let Just (ast, _) = apply com "print (1 < 1)"
                (_, out, _) = interpret (exec ast) Map.empty
             in out `shouldBe` "0\n"
        it "ltes" $
            let Just (ast, _) = apply com "print (1 <= 1)"
                (_, out, _) = interpret (exec ast) Map.empty
             in out `shouldBe` "1\n"
        it "gts" $
            let Just (ast, _) = apply com "print (1 > 1)"
                (_, out, _) = interpret (exec ast) Map.empty
             in out `shouldBe` "0\n"
        it "gtes" $
            let Just (ast, _) = apply com "print (1 >= 1)"
                (_, out, _) = interpret (exec ast) Map.empty
             in out `shouldBe` "1\n"


src/AST.hs
module AST
    ( Value
    , Identifier
    , Exp
        ( Constant
        , Variable
        , UnaryMinus
        , Plus
        , Minus
        , Times
        , Div
        , Pow
        , Complement
        , And
        , Or
        , Xor
        , Not
        , Equal
        , NotEqual
        , Less
        , LessEqual
        , Greater
        , GreaterEqual
        )
    , Com
        ( Declare
        , Assign
        , Cond
        , While
        , Seq
        , Print
        )
    ) where

        type Value = Int

        type Identifier = String

        data Exp = Constant Value
                 | Variable Identifier
                 | UnaryMinus Exp
                 | Plus Exp Exp
                 | Minus Exp Exp
                 | Times Exp Exp
                 | Div Exp Exp
                 | Pow Exp Exp
                 | Complement Exp
                 | And Exp Exp
                 | Or Exp Exp
                 | Xor Exp Exp
                 | Not Exp
                 | Equal Exp Exp
                 | NotEqual Exp Exp
                 | Less Exp Exp
                 | LessEqual Exp Exp
                 | Greater Exp Exp
                 | GreaterEqual Exp Exp
                 deriving (Eq, Show)

        data Com = Declare Identifier Exp Com
                 | Assign Identifier Exp
                 | Cond Exp Com Com
                 | While Exp Com
                 | Seq Com Com
                 | Print Exp
                 deriving (Eq, Show)


src/Parser.hs
{-# LANGUAGE InstanceSigs #-}

module Parser
    ( token
    , symbol
    , ident
    , identif
    , rexp
    , expr
    , bin
    , term
    , pow
    , factor
    , var
    , digiti
    , digit
    , relop
    , addop
    , binop
    , mulop
    , powop
    , unaryop
    , com
    , assign
    , seqv
    , cond
    , while
    , declare
    , printe
    , apply
    , applyAll
    ) where

        import Data.Char
        import Data.Maybe
        import Text.ParserCombinators.ReadP

        import AST

        token :: ReadP a -> ReadP a
        token p = do
            a <- p
            skipSpaces
            return a

        symbol :: String -> ReadP String
        symbol cs = token (string cs)

        ident :: ReadP String
        ident = do
            c <- satisfy isAlpha
            cs <- many (satisfy (\ a -> isAlpha a || isDigit a))
            return (c : cs)

        identif :: ReadP String
        identif = token ident

        rexp :: ReadP Exp
        rexp = chainl1 expr relop

        expr :: ReadP Exp
        expr = chainl1 bin addop

        bin :: ReadP Exp
        bin = chainl1 term binop

        term :: ReadP Exp
        term = chainl1 pow mulop

        pow :: ReadP Exp
        pow = chainl1 factor powop

        factor :: ReadP Exp
        factor = var +++ digiti +++ do
            symbol "("
            n <- rexp
            symbol ")"
            return n

        var :: ReadP Exp
        var = Variable <$> identif

        digiti :: ReadP Exp
        digiti = (do
            op <- unaryop
            op <$> digiti) +++ do
            p <- digit
            l <- many digit
            return (foldl (\ a b -> let Constant a' = a
                                        Constant b' = b
                                     in Constant (10 * a' + b')) (Constant 0) (p : l))

        digit :: ReadP Exp
        digit = do
            x <- token (satisfy isDigit)
            return (Constant (ord x - ord '0'))

        relop :: ReadP (Exp -> Exp -> Exp)
        relop = eq +++ ne +++ lt +++ lte +++ gt +++ gte
            where eq = do
                      symbol "="
                      return Equal
                  ne = do
                      symbol "!="
                      return NotEqual
                  lt = do
                      symbol "<"
                      return Less
                  lte = do
                      symbol "<="
                      return LessEqual
                  gt = do
                      symbol ">"
                      return Greater
                  gte = do
                      symbol ">="
                      return GreaterEqual

        addop :: ReadP (Exp -> Exp -> Exp)
        addop = plus +++ minus
            where plus = do
                      symbol "+"
                      return Plus
                  minus = do
                      symbol "-"
                      return Minus

        binop :: ReadP (Exp -> Exp -> Exp)
        binop = and +++ or +++ xor
            where and = do
                      symbol "&"
                      return And
                  or = do
                      symbol "|"
                      return Or
                  xor = do
                      symbol "^"
                      return Xor

        mulop :: ReadP (Exp -> Exp -> Exp)
        mulop = mult +++ div
            where mult = do
                      symbol "*"
                      return Times
                  div = do
                      symbol "/"
                      return Div

        powop :: ReadP (Exp -> Exp -> Exp)
        powop = do
            symbol "**"
            return Pow
        
        unaryop :: ReadP (Exp -> Exp)
        unaryop = unaryMinus +++ complement +++ not
            where unaryMinus = do
                      symbol "-"
                      return UnaryMinus
                  complement = do
                      symbol "~"
                      return Complement
                  not = do
                      symbol "!"
                      return Not

        com :: ReadP Com
        com = assign +++ seqv +++ cond +++ while +++ declare +++ printe

        assign :: ReadP Com
        assign = do
            x <- identif
            symbol ":="
            Assign x <$> rexp

        seqv :: ReadP Com
        seqv = do
            symbol "{"
            c <- com
            symbol ";" 
            d <- com
            symbol "}"
            return (Seq c d)

        cond :: ReadP Com
        cond = do
            symbol "if"
            e <- rexp
            symbol "then"
            c <- com
            symbol "else"
            Cond e c <$> com

        while :: ReadP Com
        while = do
            symbol "while"
            e <- rexp
            symbol "do"
            While e <$> com

        declare :: ReadP Com
        declare = do
            symbol "declare"
            x <- identif
            symbol "="
            e <- rexp 
            symbol "in"
            Declare x e <$> com
        
        printe :: ReadP Com
        printe = do
            symbol "print"
            Print <$> rexp

        apply :: ReadP a -> String -> Maybe (a, String)
        apply p s = toMaybe $ readP_to_S apply' s
            where toMaybe []      = Nothing
                  toMaybe (x : _) = Just x
                  apply' = do
                      skipSpaces
                      p

        applyAll :: ReadP a -> String -> [a]
        applyAll p s = let maybe = apply p s
                        in if isNothing maybe
                            then []
                            else let (a, unparsed) = fromJust maybe
                                  in if null unparsed
                                      then [a]
                                      else a : applyAll p unparsed


src/Interpreter.hs
{-# LANGUAGE InstanceSigs, TupleSections #-}

module Interpreter
    ( Interpreter
    , interpret
    , eval
    , exec
    ) where

        import Control.Applicative (Alternative, empty, (<|>))
        import Control.Monad
        import Data.Bits
        import qualified Data.Map.Strict as Map
        import Data.Maybe

        import AST
        import Lib

        type Stack = Map.Map Identifier Value
        type Output = String

        newtype Interpreter a = StateOutput {
            interpret :: Stack -> (Stack, Output, a)
        }

        instance Monad Interpreter where
            return x = StateOutput (, "", x)
            m >>= f  = StateOutput (\ s -> let (s1, o1, a1) = interpret m s
                                               (s2, o2, a2) = interpret (f a1) s1
                                            in (s2, o1 ++ o2, a2))

        instance Applicative Interpreter where
            pure  = return
            (<*>) = ap

        instance Functor Interpreter where
            fmap = liftM

        eval :: Exp -> Interpreter Value
        eval exp = case exp of
            Constant n -> return n
            Variable x -> var x
                where var id = StateOutput (\ stack -> (stack, "", fromJust (Map.lookup id stack)))
            UnaryMinus x -> do
                a <- eval x
                return (negate a)
            Plus x y -> do
                a <- eval x
                b <- eval y
                return (a + b)
            Minus x y -> do
                a <- eval x
                b <- eval y
                return (a - b)
            Times x y -> do
                a <- eval x
                b <- eval y
                return (a * b)
            Div x y -> do
                a <- eval x
                b <- eval y
                return (a `div` b)
            Pow x y -> do
                a <- eval x
                b <- eval y
                return (a ^ b)
            Complement x -> do
                a <- eval x
                return (complement a)
            And x y -> do
                a <- eval x
                b <- eval y
                return (a .&. b)
            Or x y -> do
                a <- eval x
                b <- eval y
                return (a .|. b)
            Xor x y -> do
                a <- eval x
                b <- eval y
                return (a `xor` b)
            Not x -> do
                a <- eval x
                return $ fromBool (a == 0)
            Equal x y -> do
                a <- eval x
                b <- eval y
                return $ fromBool (a == b)
            NotEqual x y -> do
                a <- eval x
                b <- eval y
                return $ fromBool (a /= b)
            Less x y -> do
                a <- eval x
                b <- eval y
                return $ fromBool (a < b)
            LessEqual x y -> do
                a <- eval x
                b <- eval y
                return $ fromBool (a <= b)
            Greater x y -> do
                a <- eval x
                b <- eval y
                return $ fromBool (a > b)
            GreaterEqual x y -> do
                a <- eval x
                b <- eval y
                return $ fromBool (a >= b)

        exec :: Com -> Interpreter () 
        exec stmt = case stmt of
            Declare id exp s -> do
                v <- eval exp
                declare id v
                exec s
                destroy id
                where declare id val = StateOutput (\ stack -> (Map.insert id val stack, "", ()))
                      destroy id = StateOutput (\ stack -> (Map.delete id stack, "", ()))
            Assign id val -> do
                v <- eval val
                assign id v
                where assign id val = StateOutput (\ stack -> (Map.adjust (const val) id stack, "", ()))
            Cond exp s1 s2 -> do
                v <- eval exp
                if v == 1
                    then exec s1
                    else exec s2
            a@(While exp s) -> do
                v <- eval exp
                if v == 0
                    then return ()
                    else do
                        exec s
                        exec a
            Print exp -> do
                v <- eval exp
                output v
                where output v = StateOutput (, show v ++ "\n", ())
            Seq s1 s2 -> do
                exec s1
                exec s2


interpreter.hs
{-# LANGUAGE InstanceSigs, LambdaCase, TupleSections #-}

import Control.Applicative (Alternative, empty, (<|>))
import Control.Monad
import Data.Char
import qualified Data.Map.Strict as Map
import Data.Maybe
import Text.ParserCombinators.ReadP

fromBool :: Bool -> Int
fromBool False = 0
fromBool True = 1

data Exp = Constant Int | Variable String | Plus Exp Exp | Minus Exp Exp | Greater Exp Exp | Less Exp Exp | Equal Exp Exp | Times Exp Exp | Div Exp Exp deriving Show

data Com = Assign String Exp | Seq Com Com | Cond Exp Com Com | While Exp Com | Declare String Exp Com | Print Exp deriving Show

type Identifier = String
type Value = Int
type Stack = Map.Map Identifier Value
type Output = String
newtype Interpreter a = StateOutput {
    interpret :: Stack -> (Stack, Output, a)
}

instance Monad Interpreter where
    return x = StateOutput (, "", x)
    m >>= f  = StateOutput (\s -> let (s1, o1, a1) = interpret m s
                                      (s2, o2, a2) = interpret (f a1) s1
                                   in (s2, o1 ++ o2, a2))

instance Applicative Interpreter where
    pure  = return
    (<*>) = ap

instance Functor Interpreter where
    fmap = liftM

eval :: Exp -> Interpreter Value
eval exp = case exp of
    Constant n -> return n
    Variable x -> var x
        where var id = StateOutput (\stack -> (stack, "", fromJust (Map.lookup id stack)))
    Plus x y -> do
        a <- eval x
        b <- eval y
        return (a + b)
    Minus x y -> do
        a <- eval x
        b <- eval y
        return (a - b)
    Times x y -> do
        a <- eval x
        b <- eval y
        return (a * b)
    Div x y -> do
        a <- eval x
        b <- eval y
        return (a `div` b)
    Less x y -> do
        a <- eval x
        b <- eval y
        return $ fromBool (a < b)
    Greater x y -> do
        a <- eval x
        b <- eval y
        return $ fromBool (a > b)
    Equal x y -> do
        a <- eval x
        b <- eval y
        return $ fromBool (a == b)

exec :: Com -> Interpreter () 
exec stmt = case stmt of
    Declare id exp s -> do
        v <- eval exp
        declare id v
        exec s
        destroy id
        where declare id val = StateOutput (\stack -> (Map.insert id val stack, "", ()))
              destroy id = StateOutput (\stack -> (Map.delete id stack, "", ()))
    Assign id val -> do
        v <- eval val
        assign id v
        where assign id val = StateOutput (\stack -> (Map.adjust (const val) id stack, "", ()))
    Cond exp s1 s2 -> do
        v <- eval exp
        if v == 1
            then exec s1
            else exec s2
    a@(While exp s) -> do
        v <- eval exp
        if v == 0
            then return ()
            else do
                exec s
                exec a
    Print exp -> do
        v <- eval exp
        output v
        where output v = StateOutput (, show v, ())
    Seq s1 s2 -> do
        exec s1
        exec s2

token :: ReadP a -> ReadP a
token p = do
    a <- p
    skipSpaces
    return a

symbol :: String -> ReadP String
symbol cs = token (string cs)

apply :: ReadP a -> String -> [(a, String)]
apply p = readP_to_S $ do
    skipSpaces
    p

ident :: ReadP String
ident = do
    c <- satisfy isAlpha
    cs <- many (satisfy (\a -> isAlpha a || isDigit a))
    return (c : cs)

identif :: ReadP String
identif = token ident

var :: ReadP Exp
var = Variable <$> identif

digit :: ReadP Exp
digit = do
    x <- token (satisfy isDigit)
    return (Constant (ord x - ord '0'))

digiti :: ReadP Exp
digiti = do
    p <- digit
    l <- many digit
    return (foldl (\a b -> let Constant a' = a
                               Constant b' = b
                            in Constant (10 * a' + b')) (Constant 0) (p : l))

rexp :: ReadP Exp
rexp = chainl1 expr relop

expr :: ReadP Exp
expr = chainl1 term addop

term :: ReadP Exp
term = chainl1 factor mulop

factor :: ReadP Exp
factor = var +++ digiti +++ do
    symbol "("
    n <- rexp
    symbol ")"
    return n

addop :: ReadP (Exp -> Exp -> Exp)
addop = do
    symbol "-"
    return Minus +++ do
        symbol "+"
        return Plus

mulop :: ReadP (Exp -> Exp -> Exp)
mulop = do
    symbol "*"
    return Times +++ do
        symbol "/"
        return Div

relop :: ReadP (Exp -> Exp -> Exp)
relop = do
    symbol ">"
    return Greater +++ do
        symbol "<"
        return Less +++ do
            symbol "="
            return Equal
  
printe :: ReadP Com
printe = do
    symbol "print"
    Print <$> rexp

assign :: ReadP Com
assign = do
    x <- identif
    symbol ":="
    Assign x <$> rexp

seqv :: ReadP Com
seqv = do
    symbol "{"
    c <- com
    symbol ";" 
    d <- com
    symbol "}"
    return (Seq c d)

cond :: ReadP Com
cond = do
    symbol "if"
    e <- rexp
    symbol "then"
    c <- com
    symbol "else"
    Cond e c <$> com

while :: ReadP Com
while = do
    symbol "while"
    e <- rexp
    symbol "do"
    While e <$> com

declare :: ReadP Com
declare = do
    symbol "declare"
    x <- identif
    symbol "="
    e <- rexp 
    symbol "in"
    Declare x e <$> com

com :: ReadP Com
com = assign +++ seqv +++ cond +++ while +++ declare +++ printe

s :: Com
s = Declare "x" (Constant 150) $
    Declare "y" (Constant 200) $
    Seq
        (While
            (Greater (Variable "x") (Constant 0))
            (Seq
                (Assign "x" (Minus (Variable "x") (Constant 1)))
                (Assign "y" (Minus (Variable "y") (Constant 1)))
            )
        )
        (Print (Variable "y"))

s' = apply com "declare x = 150 in\n declare y = 200 in\n {while x > 0 do { x:=x-1; y:=y-1 };\n print y\n }\ndeclare x = 150 in\n declare y = 200 in\n {while x > 0 do { x:=x-1; y:=y-1 };\n print y\n }\n"

test :: (Stack, Output, ())
test = interpret (exec $ fst $ head s') Map.empty
